### PR TITLE
Spec out message encryption

### PR DESCRIPTION
**Pull request by pelle**

### CODE REVIEW

File: messages/encryption.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+We currently do not support the JOSE JWT standards including the [JOSE CFRG ECDH RFC](https://tools.ietf.org/html/draft-ietf-jose-cfrg-curves-06). This is primarily for reasons of compactness. Our responses have to be limited in size to be able to comfortably fit in a QR code.
+
+We plan to support these in the future.
+
+## Encrypting the request
+
+We use the [Sealed Box](https://download.libsodium.org/doc/public-key_cryptography/sealed_boxes.html) method which is part of [libsodium](https://download.libsodium.org). This allows the recipient to decrypt a message without having to resolve the public key of the sender.
+
+A Javascript implementation can be found in [tweetnacl-sealed-box](https://github.com/whs/tweetnacl-sealed-box).
```
Comment (oed) [inline messages/encryption.md]: Seems like this lib is doing what we did before, but in a more elegant way. I'm wondering about how secure this module is. Is there a need for an audit?

File: messages/encryption.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+Create the signed [JWT payload like normal](/messages/index.md) and encrypt the resulting JWT using the `crypto_box_seal()` function from [libsodium](https://download.libsodium.org/doc/public-key_cryptography/sealed_boxes.html) or the `tweetnacl.sealedbox.open()` function from [tweetnact-sealedbox](https://github.com/whs/tweetnacl-sealed-box). The resulting data should be base64url encoded.
```
Comment (oed) [inline messages/encryption.md]: We should also do some padding of the data that is being encrypted.

File: messages/index.md [hunk 3]
```diff
@@ -56,6 +57,10 @@ These options allow you to tell the client how you want to receive the response.
 
 Each uPort compatible JWT must be signed using an secp256k1 curve. The public key is resolved for the `iss` using the [uPort PKI](/pki/index.md).
 
+### Message Encryption
+
+Some [message transports](/transports/index.md) are not secure and require messages to be encrypted. See [/messages/encryption.md] for more.
```
Comment (oed) [inline messages/index.md]: The link to encryption.md is broken here.

**Comment by christianlundkvist** [comment]
This is interesting! I didn't know about the `crypto_box_seal()` function, which basically packages up the steps of creating an ephemeral key pair to use with `crypto_box`. It is interesting how then the nonce is derived from the ephemeral public key also.

I think we should have some kind of versioning here because we will most likely need to modify this encryption format once we define the encryption layer in DIF and W3C CCG. If we just say "if it's base64 then that means it uses vanilla crypto_box_seal()" we might not be able to upgrade easily.

Also we should make sure that we are compatible with the work being done in [this MetaMask PR](https://github.com/MetaMask/eth-sig-util/pull/18) that adds encryption as a web3 layer. So that would mean either implementing `crypto_box_seal` in that PR, or modifying this one to be closer to the work there. I'm intrigued by the idea of deriving the nonce from the ephemeral pub key but that also adds complexity.

Agree with @oed regarding the need for padding. Without padding we will leak a lot of data about the payload. Not sure how to choose padding, is that something that the JWE spec deals with?

**Comment by pelle** [comment]
@oed @christianlundkvist about padding. I'm looking at approaches.

For padding value use space, `\0` or other non base64 character


- pad up to the nearest multiple of some block size
- pad with a random amount of 0-some amount

block size could be 64 bytes? But I'm open to suggestions.

Opinion? 

This is what lib sodium does: https://download.libsodium.org/doc/helpers/padding.html

**Comment by christianlundkvist** [comment]
Ok, I think this looks good! One thing here that I think we should explicitly note is we are specifically encrypting base64-encoded data in this specific use case, which is why we can pick padding data that will not be confused with message data.

I'm now thinking that padding should probably not be included in the encryption ERC since we should be able to encrypt any type of data, which makes it not really possible to know which padding data would be ok to use.

**Comment by christianlundkvist** [comment]
I also think we should make sure we are flexible in our decryption in case we want to change the padding length later (we can make sure to always strip all padding `\0` no matter how much it is)

File: messages/index.md (hunk 3 above)
Comment (pelle) [reply messages/index.md]: Ah the links have to be like that for it to work with Gatsby on the doc site

**Pull request merged**